### ISSUE TITLE
Sound desynchronization during IPS drops

### ISSUE THREAD

**Issue by Vort** [bochs-emu/Bochs#439]
While playing games, IPS may drop sometimes.
Each time it happens, sound in game becomes more and more lagged.

For testing purposes, IPS drop can be simulated by opening and closing Bochs config window.
In this video, sound is synchronized when character collects coins at 0:28.
At 0:34, I open config menu, then wait till 0:47 for IPS stabilization.
When next coins are collected at 0:49, corresponding sound appears only at 0:52.

https://github.com/user-attachments/assets/99456af8-4f92-4849-ad62-ce4d7ecbba41

I think this problem can be solved by introducing new sound configuration option, `maxlag` for example.
Which will specify maximum time in milliseconds, allowed for sound packet to wait in queue.
Packets, which are delayed for more, should be discarded.

In my opinion, this is the last annoyance, which prevents gaming with Bochs.

Test files: [superfrog.zip](https://github.com/user-attachments/files/18283610/superfrog.zip).
Version: 911e17f.

**Comment by cybersnow02** [comment]
What is the sound card emulated? Are emulated PCI devices like sound card follow the IPS / TSC timer? Everything should be sync and the sound should "stop/chop" as soon as the CPU is hanging.

**Comment by Vort** [comment]
> ... the sound should "stop/chop" as soon as the CPU is hanging

CPU emulation compensates slowdowns with subsequent speedups.
While during slowdown sound stops, during speedup extra sound data piles up, producing desync.
Dropping of such extra data should fix this issue I think.

**Comment by cybersnow02** [comment]
Yes you are right. It should discard those but in reality, on a real computer, during the speedup it should then slowdown after that, way slower, to come back to the normal situation. This is what happen on a real computer if you hear carefully when this problem happen. So the TSC Timer should follow this logic. It should desync voluntary, after the speedup, to do compensation.

So if we recapitulate:
1. High CPU demands, the emulation freeze/stop.
2. The sound "chop" but it comes back with speedup because it doesn't want to skip any data.
3. When the emulation comes back to normal, the sound is desync.

As a solution, I think we should add a second timer to bochs but this timer will be outside the emulator. We can then compare the diff between the Emulator Timer and the software/computer emulated timer. When there is a desync, Bochs will know what to do to do compensation for all kind of tasks. This could also fix alot of other desync problems.

**Comment by sskras** [comment]
> on a real computer, during the speedup it should then slowdown after that, way slower, to come back to the normal situation. This is what happen on a real computer if you hear carefully when this problem happen.

@cybersnow02, what do you mean by slowing down on a real computer?

Early computers (up to 486) had the Turbo button, pushing which made the CPU clock go 2x faster/slower.

Modern CPUs have power mgmt where the _base clock_ might drop down on idling/easy tasks, and some cores also might be pushed up to the _boost clock_ value for specific tasks (like single-threaded bursts of load).

Did you mean any of these two scenarios?

Even if so, I doubt that the latter thing (boosting/throttling) can induce sound lags in games.
I am also unsure if Bochs emulate CPUs from this era, as seemingly:

- throttling introduced in:
  * Intel Pentium 4 (2000)
  * Mobile Pentium III / Pentium M (2001–2003)
  * AMD Athlon 64 (2003)
- boosting introduced in:
  * Intel Nehalem (Turbo Boost in 2008)
  * Phenom II X6 (AMD Turbo CORE in 2010)

An explanation would be nice to have.

**Comment by cybersnow02** [comment]
@sskras I remember in 1997/1998, when I had a QDI LEGEND motherboard with a Pentium II 233mhz, the sound was always "chopping" when the BUS was overloaded. During overloaded gaming, I could experience exactly the same issue as described here. The sound was stopping, then when the computer was coming back to normal, it was playing the sound faster (sound piling up), then right after that to compensate, the sound was suddenly playing slower to sync back with the rest, in a logarithmic way. This was a strange behavior. It is hard to explain and only people that used those computers in that era might remember this behavior. The described Bochs behavior looked the same thing as what I experienced with my 90s computers on win95/win98 :)

**Comment by sskras** [comment]
@cybersnow02: thanks, fine, but IMO that's the specific software which slows down, not the hardware (which I assumed you meant originally). So the physical hardware runs without slowdowns, at a constant speeds, it only might get bottlenecked by a big computing / IO load (which is different from slowing the hardware down in real time).

So I am sure it's specific to the combination of OS and the game from these times, just not the hardware.

Now when the PC hardware gets emulated, slowdowns might mean very different things, I suppose, eg. introducing an unpleasant difference (desync) of time flow between the host and the guest. Which might be ve-e-ery different from what happens in the original, physical gear.

**Comment by Vort** [comment]
> IMO that's the specific software which slows down, not the hardware

I remember slower/faster sound playback in Bochs as well, with Wolfenstein 3D game.
It may be actually hardware feature of Sound Blaster, but I guess it happened when MIDI notes were playing.
This issue however focuses on cases where PCM sound is used.
